src/dune_rules/ctypes/ctypes_rules.ml

```diff
@@ -216,33 +216,54 @@ let build_c_program ~foreign_archives_deps ~sctx ~dir ~source_files ~scope
     in
     deps
   in
-  let build =
-    let absolute_path_hack p =
-      (* These normal path builder things construct relative paths like
-         _build/default/your/project/file.c but before dune runs gcc it actually
-         cds into _build/default, which fails, so we turn them into absolutes to
-         hack around it. *)
-      Path.relative (Path.build dir) p |> Path.to_absolute_filename
-    in
-    let action =
-      let open Action_builder.O in
-      let* include_args = Resolve.Memo.read include_args in
-      let* base_args = with_user_and_std_flags in
-      deps
-      >>> Action_builder.map cflags ~f:(fun cflags_args ->
-              let source_files = List.map source_files ~f:absolute_path_hack in
-              let output = absolute_path_hack output in
-              let args =
-                base_args @ cflags_args @ include_args @ source_files
-                @ [ "-o"; output ]
-              in
-              Action.run exe args)
-    in
-    Action_builder.with_file_targets action
-      ~file_targets:[ Path.Build.relative dir output ]
+  let action =
+    if version >= (0, 3) then
+      let args : _ Command.Args.t list =
+        [ Expand (fun ~dir:_ -> with_user_and_std_flags)
+        ; Expand (fun ~dir:_ -> cflags)
+        ; Expand (fun ~dir:_ -> Resolve.Memo.read include_args)
+        ; Deps
+            (List.map
+               ~f:(fun s -> Path.relative (Path.build dir) s)
+               source_files)
+        ; A "-o"
+        ; Target (Path.Build.relative dir output)
+        ]
+      in
+      let open Action_builder.With_targets.O in
+      Action_builder.with_no_targets deps
+      >>> Command.run ~dir:(Path.build dir) exe args
+    else
+      let build =
```

Comment:
Why do we bother with supporting old versions? The extension is unstable anyway

Reply:
I agree, that's why I want to remove it rather quickly, but I find it handy to have at least a version that supports both so that I can port existing code without having to rebuild dune at the same time. But I suppose we can shorten that by one version.